lib/ipip.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <rte_ip.h>
+#include <rte_mbuf.h>
+#include <rte_memcpy.h>
+#include <rte_byteorder.h>
+
+#include "gatekeeper_ipip.h"
+
+#define IP_VERSION              (0x40)
+/* Default IP header length == five 32-bits words. */
+#define IP_HDRLEN               (0x05)
+/* From RFC 1340. */
+#define IP_DEFTTL               (64)
+#define IP_VHL_DEF              (IP_VERSION | IP_HDRLEN)
+#define IP_DN_FRAGMENT_FLAG     (0x0040)
+
+#define IPv6_DEFAULT_VTC_FLOW   (0x60000000)
+#define IPv6_DEFAULT_HOP_LIMITS (0xFF)
+
+int
+encapsulation(struct rte_mbuf *pkt, uint8_t priority,
+	struct ipip_tunnel_info *info)
+{
+	uint64_t ol_flags = 0;
+
+	struct ether_hdr *inner_eth;
+	struct ether_hdr *outer_eth;
```

Comment:
Since there will only be one Ethernet header in the frame, the names "inner" and "outer" are a little confusing. Maybe "old_eth" and "new_eth" instead?